banditpylib/learners/collaborative_learner/collaborative_learning.py

```
@@ -0,0 +1,390 @@
+    self.__active_arms = list(range(self.__arm_num))
+
+  def __assign_arms(self, agent_ids: List[int]) ->\
+    Dict[int, List[int]]:
+    agent_arm_assignment: Dict[int, List[int]] = {}
+
+    def random_round(x: float) -> int:
+      # if x = 19.3
+      # rounded to 19 with probability 0.7
+      # rounded to 20 with probability 0.3
+      frac_x = x - int(x)
+      if np.random.uniform() > frac_x:
+        return int(x)
+      return int(x) + 1
+
+    if len(self.__active_arms) <= 1:
+      # don't assign any arms if 0/1 active arms
+      agent_arm_assignment = {}
+      for agent_id in agent_ids:
+        agent_arm_assignment[agent_id] = []
+      return agent_arm_assignment
+
+    if len(self.__active_arms) < len(agent_ids):
+      arms_assign_list = []
+      num_agents_per_arm = len(agent_ids) / len(self.__active_arms)
+      for arm in self.__active_arms[:-1]:
+        arms_assign_list += [[arm]] * \
+          random_round(num_agents_per_arm)
+      arms_assign_list += [[self.__active_arms[-1]]] * \
+        (len(agent_ids) - len(arms_assign_list))
+      random.shuffle(arms_assign_list)
+      for i, agent_id in enumerate(agent_ids):
+        agent_arm_assignment[agent_id] = arms_assign_list[i]
+    else:
+      active_arms_copy = dcopy(self.__active_arms)
+      random.shuffle(active_arms_copy)
+      for agent_id in agent_ids:
+        agent_arm_assignment[agent_id] = []
+
+      for i, arm in enumerate(active_arms_copy[:len(agent_ids)]):
+        agent_arm_assignment[agent_ids[i]].append(arm)
+      for arm in active_arms_copy[len(agent_ids):]:
+        agent_idx = random.choice(agent_ids)
+        agent_arm_assignment[agent_idx].append(arm)
+    return agent_arm_assignment
+
+  def initial_arm_assignment(self) -> Dict[int, List[int]]:
+    return self.__assign_arms(list(range(self.__num_agents)))
+
+  def elimination(self, agent_ids: List[int],
+    messages: Dict[int, Tuple[float, int]]) ->Dict[int, List[int]]:
+    if not messages:
```

When there is no message, we should raise exception.

---

It is possible that in learning, LilUCBHeuristic did not finish in horizon/2 time, which means agent return None in broadcast. So if all agents return None, then that means the algorithm failed (regret 1). Should we raise an Exception instead? What will it say - The horizon is insufficient to find the best arm ?

---

You could randomly output one arm.

---

Are you saying LilUCBHeuristic should output a random arm
Or should the elimination and arm assignment step use a random arm

---

LilUCBHeuristic should output a random arm I think

---

We could initialize `LilUCBHeursiticCollaborative` with `horizon`, and it will return the currently best arm if `horizon` is reached